### ISSUE TITLE
[Bug] RuntimeError: nms_impl: implementation for device cuda:0 not found

### ISSUE THREAD

**Issue by maxpain** [open-mmlab/mmcv#2968]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmcv/issues) and [Discussions](https://github.com/open-mmlab/mmcv/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the latest version(https://github.com/open-mmlab/mmcv).

### Environment

/bin/sh: 1: /usr/local/cuda/bin/nvcc: not found
/bin/sh: 1: x86_64-linux-gnu-gcc: not found
/bin/sh: 1: /usr/local/cuda/bin/nvcc: not found
OrderedDict([('sys.platform', 'linux'), ('Python', '3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]'), ('CUDA available', True), ('numpy_random_seed', 2147483648), ('GPU 0', 'NVIDIA L4'), ('CUDA_HOME', '/usr/local/cuda'), ('NVCC', 'Not Available'), ('GCC', 'n/a'), ('PyTorch', '2.1.0+cu121'), ('PyTorch compiling details', 'PyTorch built with:\n  - GCC 9.3\n  - C++ Version: 201703\n  - Intel(R) oneAPI Math Kernel Library Version 2022.2-Product Build 20220804 for Intel(R) 64 architecture applications\n  - Intel(R) MKL-DNN v3.1.1 (Git Hash 64f6bcbcbab628e96f33a62c3e975f8535a7bde4)\n  - OpenMP 201511 (a.k.a. OpenMP 4.5)\n  - LAPACK is enabled (usually provided by MKL)\n  - NNPACK is enabled\n  - CPU capability usage: AVX512\n  - CUDA Runtime 12.1\n  - NVCC architecture flags: -gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_90,code=sm_90\n  - CuDNN 8.9.2\n  - Magma 2.6.1\n  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=12.1, CUDNN_VERSION=8.9.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -D_GLIBCXX_USE_CXX11_ABI=0 -fabi-version=11 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-unused-private-field -Wno-aligned-allocation-unavailable -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_DISABLE_GPU_ASSERTS=ON, TORCH_VERSION=2.1.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=1, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, \n'), ('TorchVision', '0.16.0+cu121'), ('OpenCV', '4.8.1'), ('MMEngine', '0.9.0'), ('MMCV', '2.1.0'), ('MMCV Compiler', 'GCC 9.3'), ('MMCV CUDA Compiler', 'not available')])

### Reproduces the problem - code sample

```python3
from mmcv.ops import batched_nms
import torch


def check_mmcv():

    device = torch.device('cuda:0')

    bboxes = torch.randn(2, 4, device=device)
    scores = torch.randn(2, device=device)
    labels = torch.zeros(2, dtype=torch.long, device=device)
    det_bboxes, keep_idxs = batched_nms(bboxes.to(torch.float32), scores.to(torch.float32), labels, {
        'type': 'nms',
        'iou_threshold': 0.6
    })

    print('OK.')


if __name__ == '__main__':
    check_mmcv()
```

### Reproduces the problem - command or script

```shell
python test.py
```

### Reproduces the problem - error message

```
Traceback (most recent call last):
  File "/app/test.py", line 21, in <module>
    check_mmcv()
  File "/app/test.py", line 12, in check_mmcv
    det_bboxes, keep_idxs = batched_nms(bboxes.to(torch.float32), scores.to(torch.float32), labels, {
  File "/usr/local/lib/python3.10/dist-packages/mmcv/ops/nms.py", line 303, in batched_nms
    dets, keep = nms_op(boxes_for_nms, scores, **nms_cfg_)
  File "/usr/local/lib/python3.10/dist-packages/mmengine/utils/misc.py", line 395, in new_func
    output = old_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/mmcv/ops/nms.py", line 127, in nms
    inds = NMSop.apply(boxes, scores, iou_threshold, offset, score_threshold,
  File "/usr/local/lib/python3.10/dist-packages/torch/autograd/function.py", line 539, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/usr/local/lib/python3.10/dist-packages/mmcv/ops/nms.py", line 27, in forward
    inds = ext_module.nms(
RuntimeError: nms_impl: implementation for device cuda:0 not found.
```

### Additional information

I have the the problem when using mmcv==2.1.0 (installed via mim) with PyTorch 2.1.0 and CUDA 12.1 in a Dockerfile.
If I install the mmcv package from source (mim install "git+https://github.com/open-mmlab/mmcv.git@6299bc02bde35f96e0b57a6cc94ed0fda177c478") everything works fine.

Setting these ENV variables in Dockerfile doesn't help:
```Dockerfile
ENV FORCE_CUDA="1"
ENV MMCV_WITH_OPS=1
```

**Comment by maxpain** [comment]
@zhouzaida could you check, please?

**Comment by wangxinqun** [comment]
I met a similar bug as you. It turned out that I did not install the MMCV with a GPU compiler. I hope this clue will help you.

**Comment by zhouzaida** [comment]
Thanks for your feedback. I checked `https://download.openmmlab.com/mmcv/dist/cu121/torch2.1/index.html` and found some problems with cuda12.1 pre-built packages. I will try to fix it.

**Comment by zhouzaida** [comment]
Hi all, I have updated the pre-built packages for cuda12.1. Please try it again.

**Comment by maxpain** [comment]
> Hi all, I have updated the pre-built packages for cuda12.1. Please try it again.
> 
> ```shell
> pip install -U openmim
> mim install mmcv
> ```

It's working now, thanks!

**Comment by wonjun-dev** [comment]
Which version of mmcv is working? I have same problem...

**Comment by wonjun-dev** [comment]
I used mmcv==2.1.0, but it still raised same error.
This is my settings:
pytorch==2.1.0a0+32f93b1
cuda==12.2
mmcv==2.1.0
mmdet==3.2.0
mmengine==0.10.2
mmpretrain==1.1.1

**Comment by liuhongji2000** [comment]
@wonjun-dev try cuda12.1?

**Comment by wonjun-dev** [comment]
> @wonjun-dev try cuda12.1?

I tried it, but it raised different error:

ImportError: /usr/local/lib/python3.10/dist-packages/mmcv/_ext.cpython-310-x86_64-linux-gnu.so: undefined symbol: _ZN3c105ErrorC2ENS_14SourceLocationESs

**Comment by rvandeghen** [comment]
Hi @zhouzaida,

Do you have any idea if this bug still occurs with `torch==2.2.1` ?

Here is the output of `collect_env`:

```
sys.platform: linux
Python: 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:53:32) [GCC 12.3.0]
CUDA available: True
MUSA available: False
numpy_random_seed: 2147483648
GPU 0: NVIDIA A100-SXM4-40GB
CUDA_HOME: None
GCC: gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-10)
PyTorch: 2.2.1+cu121
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201703
  - Intel(R) oneAPI Math Kernel Library Version 2022.2-Product Build 20220804 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v3.3.2 (Git Hash 2dc95a2ad0841e29db8b22fbccaf3e5da7992b01)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 12.1
  - NVCC architecture flags: -gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_90,code=sm_90
  - CuDNN 8.9.2
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=12.1, CUDNN_VERSION=8.9.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -D_GLIBCXX_USE_CXX11_ABI=0 -fabi-version=11 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=pedantic -Wno-error=old-style-cast -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=2.2.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=1, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, USE_ROCM_KERNEL_ASSERT=OFF, 

TorchVision: 0.17.1+cu121
OpenCV: 4.9.0
MMEngine: 0.10.3
MMDetection: 3.3.0+cfd5d3a
```

**Comment by vedrusss** [comment]
I've found the solution how to build mmcv with gpu nms_ops support while building docker image.
Actually, it's not a problem to build-install it manually within running docker container - just force re-build it there. But the problem was to do it while building docker image.
Commands like

    mim uninstall mmcv
    mim install mmcv==2.1.0 

helped within running docker container but had no effect within Dockerfile (in RUN lines) - it re-built mmcv which still didn't work.
The reason of that - within Dockerfile it built mmcv without cuda support (at least, for nms_ops). To force it I had to pass ENV values:

    ENV FORCE_CUDA="1" \
        MMCV_WITH_OPS=1

But `FORCE_CUDA="1"` leaded to crash with error 

    arch_list[-1] += '+PTX'
    IndexError: list index out of range

To prevent that I had to specify my GPU compute capability. One may either specify a wide range by

    ENV TORCH_CUDA_ARCH_LIST="5.2 6.0 6.1 7.0 7.5 8.0 8.6 9.0+PTX"

which will lead to super long building time or specify exact value obtained by command `nvidia-smi --query-gpu=compute_cap --format=csv,noheader`
Unfortunately there is no way how to put such command in Dockerfile ENV line. My solution was to run it within `docker build` command providing that way argument to the builder. Next I used it within Dockerfile like

    ARG gpu_comp
    ENV TORCH_CUDA_ARCH_LIST="${gpu_comp}+PTX"

This way docker image was built with correct mmcv.
